File: csharp_src/Runtime/Tsv.cs
// Tsv - one place for the tab-separated line handling the oracle, the dispatcher, and the propose
// flow all share: CRLF-tolerant splitting into non-empty lines (and into tab-split rows).

using System.Collections.Generic;

namespace Icm
{
    internal static class Tsv
    {
        // Non-empty lines, with trailing '\r' stripped (tolerate Windows CRLF).
        public static List<string> NonEmptyLines(string text)
        {
            var outl = new List<string>();
            foreach (string raw in text.Split('\n'))
            {
                string line = raw.TrimEnd('\r');
                if (line.Trim().Length != 0) outl.Add(line);
            }
            return outl;
        }

        // Non-empty lines, each split into cells on tabs.
        public static List<string[]> Rows(string text)
        {
            var outl = new List<string[]>();
            foreach (string line in NonEmptyLines(text)) outl.Add(line.Split('\t'));
            return outl;
        }
    }
}

File: go_src/internal/oracle/tsv.go
// Tsv - one place for the tab-separated line handling the oracle, the dispatcher, and the propose
// flow share: CRLF-tolerant splitting into non-empty lines (and into tab-split rows). Port of
// src.bak/Runtime/Tsv.cs.
package oracle

import "strings"

// NonEmptyLines returns the non-empty lines, with a trailing '\r' stripped (tolerate Windows CRLF).
func NonEmptyLines(text string) []string {
	var out []string
	for _, raw := range strings.Split(text, "\n") {
		line := strings.TrimRight(raw, "\r")
		if strings.TrimSpace(line) != "" {
			out = append(out, line)
		}
	}
	return out
}

// Rows returns the non-empty lines, each split into cells on tabs.
func Rows(text string) [][]string {
	var out [][]string
	for _, line := range NonEmptyLines(text) {
		out = append(out, strings.Split(line, "\t"))
	}
	return out
}
